### DISCUSSION TITLE
What is the difference between trace and print methods?

### DISCUSSION THREAD

**Discussion by usmanovbf** [open-policy-agent/community#646] [OPA and Rego]
this one https://www.openpolicyagent.org/docs/latest/policy-reference/#tracing and this https://www.openpolicyagent.org/docs/latest/policy-reference/#debugging

**Comment by charlieegan3** [comment]
Hey, sorry for the delay here ❄️ 

Trace is only shown when using explain, but notes are shown relative to where they're called. Print 'just' prints to the console during execution.

Print is newer, and generally speaking, easier to use for debugging.

```
$ opa eval --explain=notes --format=pretty 'trace("hello world")'
query:1     Enter trace("hello world", _)
query:1     | Note "hello world"
true

$ opa eval 'print("hello world")'
hello world
{
  "result": [
    {
      "expressions": [
        {
          "value": true,
          "text": "print(\"hello world\")",
          "location": {
            "row": 1,
            "col": 1
          }
        }
      ]
    }
  ]
}
```
